Boost dependency can be (mostly) upgraded

Previously, there was an incompatibility between Boost's `noinline` and what was provided within the HIP compiler which was initially reported in #658.  Since then, the Boost library has been patched through boostorg/config#393 which was included in Boost version 1.78.  As a result, I have tried again to see if I can avoid having to recompile Boost and use the system version and it mostly works.

In particular, the compilation works fine for all files except for the following 9 files:
- `src/find_db.cpp`
- `src/binary_cache.cpp`
- `src/sqlite_db.cpp`
- `src/solver/conv_asm_3x3u.cpp`
- `src/ocl/gcn_asm_utils.cpp`
- `src/ocl/mloNeuron.cpp`
- `src/ocl/mloPooling.cpp`
- `src/ocl/mloNorm.cpp`
- `src/hip/hip_build_utils.cpp`

The error message is similar to the original issue:

```text
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/find_db.cpp:29:
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/include/miopen/handle.hpp:30:
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/include/miopen/kernel_info.hpp:33:
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/include/miopen/kernel.hpp:53:
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/include/miopen/hipoc_kernel.hpp:34:
In file included from /tmp/makepkg/miopen-hip/src/MIOpen-rocm-5.0.2/src/include/miopen/op_kernel_args.hpp:8:
In file included from /usr/include/boost/container/small_vector.hpp:27:
/usr/include/boost/container/vector.hpp:2754:4: error: use of undeclared identifier 'noinline'; did you mean 'inline'?
   BOOST_CONTAINER_NOINLINE iterator priv_insert_forward_range_no_capacity
   ^
/usr/include/boost/container/detail/workaround.hpp:116:37: note: expanded from macro 'BOOST_CONTAINER_NOINLINE'
   #define BOOST_CONTAINER_NOINLINE BOOST_NOINLINE
                                    ^
/usr/include/boost/config/detail/suffix.hpp:639:46: note: expanded from macro 'BOOST_NOINLINE'
#      define BOOST_NOINLINE __attribute__ ((__noinline__))
                                             ^
/opt/rocm/hip/../include/hip/amd_detail/host_defines.h:50:37: note: expanded from macro '__noinline__'
#define __noinline__ __attribute__((noinline))
                                    ^
```

and in all of the above cases, the addition of `#include <hip/hip_version.h>` as an additional header in each `.cpp` file with no other change to the code fixes the compilation error.  I did also try adding this header to `hipoc_kernel.hpp` but that did not resolve the issue.

I'm happy to create a PR to address this issue, though I would like to understand why adding the missing `hip_version.h` header to `hipoc_kernel.hpp` does not appear to resolve the issue.

I have checked this against the latest develop commit, and the issue (at least when compiling `find_db.cpp`) is still there.

--- a/include/miopen/config.h.in
+++ b/include/miopen/config.h.in
@@ -118,3 +118,9 @@
      HIP_PACKAGE_VERSION_PATCH)
 
 #endif
+
+/// Workaround for https://github.com/boostorg/config/issues/392#issuecomment-1109889533
+/// TODO: Remove the W/A as soon we switch to the properly fixed boost.
+#if MIOPEN_BACKEND_HIP
+#include <hip/hip_version.h>
+#endif

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,5 +1,5 @@
 RadeonOpenCompute/rocm-cmake@839cf61c355f997283e3baaf245fa9acf2e3baa0 --build
 sqlite3@3.17 -DCMAKE_POSITION_INDEPENDENT_CODE=On
-boost@1.72 -DCMAKE_POSITION_INDEPENDENT_CODE=On --build
+boost@1.79 -DCMAKE_POSITION_INDEPENDENT_CODE=On --build
 half,https://github.com/pfultz2/half/archive/1.12.0.tar.gz -X header -H sha256:0a08660b68abb176ebc2a0cdf8de46e3182a7f46c66443bb80dbfaaec98cf969 --build
 ROCmSoftwarePlatform/llvm-project-mlir@release/rocm-5.2 -DBUILD_FAT_LIBMLIRMIOPEN=1

--- a/src/exec_utils.cpp
+++ b/src/exec_utils.cpp
@@ -30,7 +30,6 @@
 #include <istream>
 #include <ostream>
 #include <string>
-#include <cstdio>
 #include <array>
 #include <cassert>
 

--- a/src/execution_context.cpp
+++ b/src/execution_context.cpp
@@ -185,6 +185,7 @@ static bool IsAmdRocmOpencl(miopen::ExecutionContext& context)
 #else
         true;
 #endif // MIOPEN_BACKEND_OPENCL
+    // cppcheck-suppress knownConditionTrueFalse
     if(ret_bool)
     {
         static const rocm_meta_version ret_rmv = AmdRocmMetadataVersionDetect(context);

--- a/src/gemm_v2.cpp
+++ b/src/gemm_v2.cpp
@@ -40,7 +40,7 @@
 
 #if MIOPEN_USE_ROCBLAS
 #include <half.hpp>
-#include <rocblas.h>
+#include <rocblas/rocblas.h>
 #include <miopen/perf_field.hpp>
 #endif
 

--- a/src/gemm_v2.cpp
+++ b/src/gemm_v2.cpp
@@ -40,7 +40,11 @@
 
 #if MIOPEN_USE_ROCBLAS
 #include <half.hpp>
+#if HIP_PACKAGE_VERSION_FLAT <= 5001999999ULL
+#include <rocblas.h>
+#else
 #include <rocblas/rocblas.h>
+#endif
 #include <miopen/perf_field.hpp>
 #endif
 

--- a/src/hip/handlehip.cpp
+++ b/src/hip/handlehip.cpp
@@ -45,9 +45,7 @@
 #endif
 
 #include <boost/filesystem.hpp>
-#include <miopen/handle_lock.hpp>
 #include <miopen/load_file.hpp>
-#include <miopen/gemm_geometry.hpp>
 
 #ifndef _WIN32
 #include <unistd.h>

--- a/src/hip/hip_build_utils.cpp
+++ b/src/hip/hip_build_utils.cpp
@@ -57,7 +57,7 @@ static boost::filesystem::path HipBuildImpl(boost::optional<TmpDir>& tmp_dir,
         auto inc_list = GetHipKernelIncList();
         auto inc_path = tmp_dir->path;
         boost::filesystem::create_directories(inc_path);
-        for(auto inc_file : inc_list)
+        for(const auto& inc_file : inc_list)
         {
             auto inc_src = GetKernelInc(inc_file);
             WriteFile(inc_src, inc_path / inc_file);

--- a/src/hipoc/hipoc_program.cpp
+++ b/src/hipoc/hipoc_program.cpp
@@ -39,7 +39,6 @@
 #include <miopen/write_file.hpp>
 #include <miopen/env.hpp>
 #include <miopen/comgr.hpp>
-#include <miopen/logger.hpp>
 #include <boost/optional.hpp>
 
 #include <cstring>

--- a/src/include/miopen/find_db.hpp
+++ b/src/include/miopen/find_db.hpp
@@ -27,6 +27,7 @@
 #ifndef GUARD_MIOPEN_FIND_DB_HPP_
 #define GUARD_MIOPEN_FIND_DB_HPP_
 
+#include <miopen/config.h>
 #include <miopen/db.hpp>
 #include <miopen/db_path.hpp>
 #include <miopen/db_record.hpp>

--- a/src/include/miopen/handle.hpp
+++ b/src/include/miopen/handle.hpp
@@ -52,7 +52,7 @@
 
 #if MIOPEN_USE_ROCBLAS
 #include <miopen/manage_ptr.hpp>
-#include <rocblas.h>
+#include <rocblas/rocblas.h>
 #endif
 
 namespace miopen {

--- a/src/include/miopen/handle.hpp
+++ b/src/include/miopen/handle.hpp
@@ -52,7 +52,11 @@
 
 #if MIOPEN_USE_ROCBLAS
 #include <miopen/manage_ptr.hpp>
+#if HIP_PACKAGE_VERSION_FLAT <= 5001999999ULL
+#include <rocblas.h>
+#else
 #include <rocblas/rocblas.h>
+#endif
 #endif
 
 namespace miopen {

--- a/src/include/miopen/mdg_expr.hpp
+++ b/src/include/miopen/mdg_expr.hpp
@@ -16,7 +16,7 @@
 #define BOOST_FOREACH(x, y) for(x : y) // NOLINT
 #endif
 #include <boost/spirit/include/qi.hpp>
-#include <boost/spirit/include/phoenix.hpp>
+#include <boost/phoenix.hpp>
 #include <boost/fusion/adapted.hpp>
 #include <boost/spirit/include/support_utree.hpp>
 #include <miopen/fusion_ops.hpp>

--- a/src/include/miopen/mlo_internal.hpp
+++ b/src/include/miopen/mlo_internal.hpp
@@ -98,14 +98,8 @@ POSSIBILITY OF SUCH DAMAGE.
 
 #include <iostream>
 #include <fstream>
-#include <iomanip>
 #include <sstream>
-#include <string>
 #include <ctime>
-#include <cmath>
-#include <ctime>
-#include <cstdlib>
-#include <cstdio>
 #include <cstring>
 #include <vector>
 #include <numeric>

--- a/src/include/miopen/op_kernel_args.hpp
+++ b/src/include/miopen/op_kernel_args.hpp
@@ -32,7 +32,7 @@ struct OpKernelArg
 
     std::size_t size() const { return buffer.size(); };
 #if HIP_PACKAGE_VERSION_FLAT <= 5001999999ULL
-    std::vector<char, 8> buffer;
+    std::vector<char> buffer;
 #else
     boost::container::small_vector<char, 8> buffer;
 #endif

--- a/src/include/miopen/op_kernel_args.hpp
+++ b/src/include/miopen/op_kernel_args.hpp
@@ -4,8 +4,11 @@
 #include <type_traits>
 #include <cstdint>
 #include <half.hpp>
+#if HIP_PACKAGE_VERSION_FLAT > 5001999999ULL
 #include <miopen/config.h>
 #include <boost/container/small_vector.hpp>
+#endif
+
 struct OpKernelArg
 {
 
@@ -28,7 +31,11 @@ struct OpKernelArg
     }
 
     std::size_t size() const { return buffer.size(); };
+#if HIP_PACKAGE_VERSION_FLAT <= 5001999999ULL
+    std::vector<char, 8> buffer;
+#else
     boost::container::small_vector<char, 8> buffer;
+#endif
     bool is_ptr = false;
 };
 

--- a/src/include/miopen/op_kernel_args.hpp
+++ b/src/include/miopen/op_kernel_args.hpp
@@ -4,7 +4,7 @@
 #include <type_traits>
 #include <cstdint>
 #include <half.hpp>
-
+#include <miopen/config.h>
 #include <boost/container/small_vector.hpp>
 struct OpKernelArg
 {

--- a/src/mlo_dir_conv.cpp
+++ b/src/mlo_dir_conv.cpp
@@ -48,15 +48,9 @@
 #include <sstream>
 #include <unordered_map>
 
-#include <miopen/solver.hpp>
 #if MIOPEN_ENABLE_SQLITE
 #include <miopen/sqlite_db.hpp>
 #endif
-#include <miopen/db.hpp>
-#include <miopen/env.hpp>
-#include <miopen/gcn_asm_utils.hpp>
-#include <miopen/mlo_internal.hpp>
-#include <miopen/mlo_utils.hpp>
 
 // Only select the first applicable igemm solver due to long compilation time
 // (JIRA SWDEV-227826)

--- a/src/ocl/convolutionocl.cpp
+++ b/src/ocl/convolutionocl.cpp
@@ -28,7 +28,6 @@
 #include <miopen/check_numerics.hpp>
 #include <miopen/config.h>
 #include <miopen/convolution.hpp>
-#include <miopen/conv_algo_name.hpp>
 #include <miopen/db.hpp>
 #include <miopen/db_record.hpp>
 #include <miopen/env.hpp>

--- a/src/ocl/fusionopbiasbnactivocl.cpp
+++ b/src/ocl/fusionopbiasbnactivocl.cpp
@@ -619,7 +619,7 @@ miopenStatus_t BatchNormFwdTrainFusionOpDescriptor::GetCompileParms(
     {
         add += " -DMIOPEN_USE_FPMIX=1";
     }
-    
+
     add += " -DMIO_BN_N=" + std::to_string(n) + " -DMIO_BN_C=" + std::to_string(c) +
            " -DMIO_BN_HW=" + std::to_string(in_cstride) +
            " -DMIO_BN_NHW=" + std::to_string(n * h * w) +

--- a/src/ocl/fusionopbiasbnactivocl.cpp
+++ b/src/ocl/fusionopbiasbnactivocl.cpp
@@ -619,7 +619,7 @@ miopenStatus_t BatchNormFwdTrainFusionOpDescriptor::GetCompileParms(
     {
         add += " -DMIOPEN_USE_FPMIX=1";
     }
-
+    
     add += " -DMIO_BN_N=" + std::to_string(n) + " -DMIO_BN_C=" + std::to_string(c) +
            " -DMIO_BN_HW=" + std::to_string(in_cstride) +
            " -DMIO_BN_NHW=" + std::to_string(n * h * w) +
@@ -631,6 +631,7 @@ miopenStatus_t BatchNormFwdTrainFusionOpDescriptor::GetCompileParms(
            " -DMIO_BN_LDS_SIZE=" + std::to_string(ldsnogcn) +
            " -DMIO_BN_LDSGCN_SIZE=" + std::to_string(ldsgcn) +
            " -DMIOPEN_READ_UNIT=" + std::to_string(read_unit) + " -DMIOPEN_READ_TYPE=" + READ_TYPE +
+           // cppcheck-suppress knownConditionTrueFalse
            " -DMIO_SAVE_MEAN_VARIANCE=" + (saveBatchStats ? "1" : "0") +
            " -DMIO_RUNNING_RESULT=" + ((savePopStats) ? "1" : "0") +
            " -DMIO_BN_VARIANT=" + std::to_string(variant) +

--- a/src/pooling.cpp
+++ b/src/pooling.cpp
@@ -32,7 +32,6 @@
 
 #include <cassert>
 #include <cmath>
-#include <cmath>
 
 namespace miopen {
 

--- a/src/readonlyramdb.cpp
+++ b/src/readonlyramdb.cpp
@@ -134,6 +134,7 @@ void ReadonlyRamDb::Prefetch(bool warn_if_unreadable)
         if(db_path.empty())
             return;
         constexpr bool isEmbedded = MIOPEN_EMBED_DB;
+        // cppcheck-suppress knownConditionTrueFalse
         if(!debug::rordb_embed_fs_override() && isEmbedded)
         {
 #if MIOPEN_EMBED_DB

--- a/src/softmax_api.cpp
+++ b/src/softmax_api.cpp
@@ -27,7 +27,6 @@
 #include <miopen/errors.hpp>
 #include <miopen/handle.hpp>
 #include <miopen/logger.hpp>
-#include <miopen/handle.hpp>
 #include <miopen/tensor_ops.hpp>
 
 static void LogCmdSoftmax(const miopenTensorDescriptor_t xDesc,

--- a/src/solver/conv_bin_winoRxS.cpp
+++ b/src/solver/conv_bin_winoRxS.cpp
@@ -30,7 +30,6 @@
 #include <miopen/conv/wrw_invoke_params.hpp>
 #include <miopen/env.hpp>
 #include <miopen/kernel_build_params.hpp>
-#include <miopen/conv/compiled_in_parameters.hpp>
 #include <miopen/conv/data_invoke_params.hpp>
 #include <miopen/conv/tensors.hpp>
 

--- a/src/solver/conv_ocl_dir2D11x11.cpp
+++ b/src/solver/conv_ocl_dir2D11x11.cpp
@@ -133,6 +133,7 @@ ConvSolution ConvOclDirectFwd11x11::GetSolution(const ConvolutionContext& params
     // this one is valid only till _FLOAT8
     // but it's not an error, the kernel does not use these types at all
     static const std::string READ_TYPE =
+        // cppcheck-suppress knownConditionTrueFalse
         (read_unit == 1) ? "_FLOAT" : "_FLOAT" + std::to_string((read_unit));
 
     // param

